### PR TITLE
Plausibility test bom refs

### LINKED ISSUE #5: Add a plausibility check for bom-refs
This would be used for e.g. finding orphaned bom-refs, or selfdependence in the dependencies.

### PR DESCRIPTION
**Pull request by CBeck-96**
closes #5

### CODE REVIEW

File: cdxev/plausibility_check.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+    for current_reference in list_of_actual_bom_refs:
+        list_of_upstream_references = get_upstream_dependency_bom_refs(
+            current_reference, sbom.get("dependencies", [])
+        )
+        if current_reference in list_of_upstream_references:
+            errors.append(create_error_circular_reference(current_reference, sbom))
+
+        # check if the dependency tree is connected, i.e. that the product
+        # decribed by the sbom depends directly or indirectly on every component.
+        # also checks that every component is depended on
```
Comment (mmarseu) [inline cdxev/plausibility_check.py:142]: This is a misinterpretation of CycloneDX. Not all components in an SBOM are required to be part of the same dependency tree.

That is why *dependencies* and *assemblies* are both a thing in CycloneDX. There are products assembled from multiple components, where there is no *depends-on* relationship but rather an *is-part-of* relationship.

**Review by mmarseu — CHANGES_REQUESTED** [review]
Shouldn't this check be part of the *validate* command? It is certainly a type of validation.  
If it remains a separate command, `available_commands.md` should be updated.

IMO, the check that all bom-refs are unique is missing. Your code depends on this assumption but doesn't make sure it is correct. Again, this would be better suited for the *validate* command, because it is a requirement in the spec. It just can't be expressed in JSON schema.

**Comment by italvi** [comment]
> Shouldn't this check be part of the _validate_ command? It is certainly a type of validation. If it remains a separate command, `available_commands.md` should be updated.
> 
> IMO, the check that all bom-refs are unique is missing. Your code depends on this assumption but doesn't make sure it is correct. Again, this would be better suited for the _validate_ command, because it is a requirement in the spec. It just can't be expressed in JSON schema.

Agree, though this should not be a default check when executing validate and be controlled through a flag. 

Regarding the uniqueness of `bom-ref` within **whole** (for `dependsOn` the `uniqueItems` property is already used) SBOM you are correct, seems like `jsonschema` does not support unique key-value pairs, [yet](https://github.com/python-jsonschema/check-jsonschema/issues/60). But easy to loop through the SBOM and do this check ourselves.

File: cdxev/__main__.py [hunk 2]
```diff
@@ -297,6 +298,16 @@ def create_validation_parser(
...
         type=str,
     )
+    parser.add_argument(
+        "--plausability-check",
+        metavar="<plausability-check>",
+        choices=["yes", "y"],
+        help=(
+            "y/yes if the plausibility of the bom-refs in the"
+            "sbom should also be checked"
+        ),
```
Comment (italvi) [inline cdxev/__main__.py:308]: Why this additional choice? Why not providing the flag `plausibility-check` means `true`?

**Review by italvi — CHANGES_REQUESTED** [review]

File: tests/test_validate.py [hunk 3]
```diff
@@ -13,6 +20,10 @@
     path_to_folder_with_test_sboms + "Acme_Application_9.1.1_20220217T101458.cdx.json"
 )
 
+path_to_second_sbom = (
+    path_to_folder_with_test_sboms + "sub_programm_T5.0.3.96_20220217T101458_cdx.json"
+)
```
[the remark this answers is not in the document]
Comment (chbk96) [reply tests/test_validate.py:25]: adapting the tests for plausibility to this one is, and in doing so changing it to fit the requirements, would be work, that seemed unnecessary

[the remark this answers is not in the document]
Comment (chbk96) [reply tests/test_validate.py]: removed

**Comment by mmarseu** [comment]
Has this PR been abandoned? It's been in my todo list forever but now I'm not even sure there is a point in reviewing this anymore.

**Comment by CBeck-96** [comment]
> Has this PR been abandoned? It's been in my todo list forever but now I'm not even sure there is a point in reviewing this anymore.

Considering there are other priorities and demand seems low, i will close it for the time beeing and tackle the subject at a more oppurtune time (propably version 1.0 or higher).

**Comment by CBeck-96** [comment]
Considering there are other priorities and demand seems low, i will close it for the time beeing and tackle the subject at a more oppurtune time (propably version 1.0 or higher).

**Pull request closed without merging**